syntax = "proto3";

package bazaar;

service BazaarService {
    rpc ElectionMessage (ElectionRequest) returns (ElectionResponse);
    rpc AnnounceLeader (LeaderAnnouncement) returns (LeaderResponse);
    rpc RegisterProduct (ProductDetails) returns (RegisterResponse);
    rpc BuyRequest(BuyRequestMessage) returns (BuyReturnResponse);
    rpc ClockUpdate(ClockMessage) returns (ClockUpdateResponse);
    rpc PurchaseProcessed(PurchaseMessage) returns (PurchaseResponse);
    
    rpc RegistrationProcessed(RegisterResponse) returns (AckMessage);
    
    rpc WarehouseCommunicationBuyer(WCBMessage) returns (PurchaseMessage);
    rpc WarehouseCommunicationSeller(WCSMessage) returns (RegisterResponse);
    rpc HeartBeat(PingMessage) returns (PingMessage);
    rpc TraderFailure(FailedTraderMessage) returns (AckMessage);
    rpc SyncCache(CacheState) returns (AckMessage); 
   
}


message CacheState
{
    int32 boar=1;
    int32 fish=2;
    int32 salt=3;
}
message FailedTraderMessage
{
    int32 trader_id=1;
    string message=2;
}
message AckMessage
{
    string message=1;
}
message PingMessage
{
    string message=1;
}
message WCBMessage
{   int32 buyer_id=1; 
    string product=2;
    int32 quantity=3;
    int32 request_no=4;
    int32 trader_id=5;
    string status=6; 

}
message WCSMessage
{
    int32 seller_id = 1;
    string product = 2;
    int32 quantity = 3; 
    int32 registration_no=4;
    int32 trader_id=5; 
}
 

message ClockMessage
{
    int32 clock_value=1;
}
message ClockUpdateResponse
{
    string message=1;
}
message BuyRequestMessage {
    int32 buyer_id=1; 
    string product=2;
    int32 quantity=3;
    int32 clock=4;
    int32 request_no=5;
}
message BuyReturnResponse{ 
    string message=1;

}
message PurchaseMessage{
    string message=1; 
    int32 buyer_id=2; 
    string product=3;
    int32 quantity=4;
    int32 request_no=5; 

}
message PurchaseResponse
{
    string message=1;
}
message TransactionMessage {
    int32 buyer_id=1;
    string product=2; 
    int32 quantity=3;
    float amount_credited=4;
    string out_of_stock=5;
    int32 request_no=6;
} 
message TransactionResponse{
    string message=1;
    
}
message ElectionRequest {
    int32 node_id = 1;
    int32 sender_id = 2;
}

message ElectionResponse {
    bool acknowledgment = 1;
}

message LeaderAnnouncement {
    int32 leader_id = 1;
}

message LeaderResponse {
    string message = 1;
   
}

message ProductDetails {
    int32 seller_id = 1;
    string product = 2;
    int32 quantity = 3;
    int32 registration_no=4; 
    int32 clock=5; 
}

message RegisterResponse { 
    int32 seller_id = 1;
    string product = 2;
    int32 quantity = 3;
    int32 registration_no=4; 
    float amount_credited= 5; 
    string message = 6;
}

